### PR TITLE
[VKCI-280] Add functionality in gateway.go to report if it is using Ip Spaces or not

### PR DESCRIPTION
**Pull request by rocknes**
Added method IsUsingIpSpaces in gateway.go

Testing Done:
Added test : TestGatewayUsingIpSpaces in ip_spaces_test.go

- - -
This change is [Reviewable](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/341)

### CODE REVIEW

File: pkg/vcdsdk/gateway.go [hunk 1]
```diff
@@ -1455,16 +1455,35 @@ func (gatewayManager *GatewayManager) GetLoadBalancer(ctx context.Context, virtu
...
+		(gm.NetworkBackingType == swaggerClient.NSXT_FIXED_SEGMENT_BackingNetworkType) ||
+			(gm.NetworkBackingType == swaggerClient.NSXT_FLEXIBLE_SEGMENT_BackingNetworkType)
 
 	return isNSXTBackedGateway
 }
 
-func (gatewayManager *GatewayManager) GetLoadBalancerPool(ctx context.Context, lbPoolName string) (*swaggerClient.EntityReference, error) {
-	lbPoolRef, err := gatewayManager.getLoadBalancerPool(ctx, lbPoolName)
+func (gm *GatewayManager) IsUsingIpSpaces() (bool, error) {
+	vdc := gm.Client.VDC
```
Comment (arunmk) [inline pkg/vcdsdk/gateway.go:1467]: We may need to do some nil checks here for the inner fields.

File: pkg/vcdsdk/gateway.go [hunk 2]
```diff
@@ -1455,16 +1455,35 @@ func (gatewayManager *GatewayManager) GetLoadBalancer(ctx context.Context, virtu
...
+	if err != nil {
+		return false, err
+	}
+
+	edgeGatewayUplinks := edgeGateway.EdgeGateway.EdgeGatewayUplinks
+	if len(edgeGatewayUplinks) != 1 {
+		return false, fmt.Errorf("found Multiple uplinks for gateway `%s`, expecting 1", edgeGatewayName)
+	}
+
+	return *edgeGatewayUplinks[0].UsingIpSpace, nil
```
Comment (arunmk) [inline pkg/vcdsdk/gateway.go:1482]: A nil check for `edgeGatewayUplinks[0]` may also be needed.

**Review by arunmk — APPROVED** [review]
Looks good to me. Some basic nits, because I am not sure how things would work when this is tested against a non-ip-spaces gateway. Please do test it once that way as well.

**Review by lzichong — APPROVED** [review]
:lgtm: after addressing comments

> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/341#-:-NpaIsXd8UMKnmivSigh:b-98fkod)* status: all files reviewed, 2 unresolved discussions (waiting on @Anirudh9794 and @arunmk)

___
*[`pkg/vcdsdk/gateway.go` line 1467 at r1](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/341#gh-1474744634:-NpaEgd-BlmFcxVweVnv:b-h6nou6) ([raw file](https://github.com/vmware/cloud-provider-for-cloud-director/blob/48b87056c556185a499875dc21f86e0e5c1d4ecd/pkg/vcdsdk/gateway.go#L1467)):*
<details><summary><i>Previously, arunmk (Arun M. Krishnakumar) wrote…</i></summary><blockquote>

We may need to do some nil checks here for the inner fields.
</blockquote></details>

+1 for vdc nil check
___
*[`pkg/vcdsdk/gateway.go` line 1482 at r1](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/341#gh-1474752127:-NpaIWxn21viePbNKhbZ:b-uuhcyv) ([raw file](https://github.com/vmware/cloud-provider-for-cloud-director/blob/48b87056c556185a499875dc21f86e0e5c1d4ecd/pkg/vcdsdk/gateway.go#L1482)):*
<details><summary><i>Previously, arunmk (Arun M. Krishnakumar) wrote…</i></summary><blockquote>

A nil check for `edgeGatewayUplinks[0]` may also be needed.
</blockquote></details>

I think we will also need a nil check for `UsingIpSpace` as it is `*bool`, this may cause a NPE if using ip space is not found as part of the RDE response due to it being omitempty

**Review by rocknes — COMMENTED** [review]
I have already added a test, to test the code against gateway using IP blocks :)

> *[Reviewable](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/341#-:-NpaQBzu0caDSiz7ZBmX:bb8ou2h)* status: all files reviewed, 2 unresolved discussions (waiting on @Anirudh9794 and @arunmk)

**Review by rocknes — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/341)* status: 3 of 4 files reviewed, 2 unresolved discussions (waiting on @Anirudh9794, @arunmk, and @lzichong)

___
*[`pkg/vcdsdk/gateway.go` line 1467 at r1](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/341#gh-1474744634:-NparVBMBHHiApWwy306:bx3iq6m) ([raw file](https://github.com/vmware/cloud-provider-for-cloud-director/blob/48b87056c556185a499875dc21f86e0e5c1d4ecd/pkg/vcdsdk/gateway.go#L1467)):*
<details><summary><i>Previously, lzichong wrote…</i></summary><blockquote>

+1 for vdc nil check
</blockquote></details>

Addressed
___
*[`pkg/vcdsdk/gateway.go` line 1482 at r1](https://reviewable.io/reviews/vmware/cloud-provider-for-cloud-director/341#gh-1474752127:-NparXyC3oxmbz5FeU6e:b-lm8z72) ([raw file](https://github.com/vmware/cloud-provider-for-cloud-director/blob/48b87056c556185a499875dc21f86e0e5c1d4ecd/pkg/vcdsdk/gateway.go#L1482)):*
<details><summary><i>Previously, lzichong wrote…</i></summary><blockquote>

I think we will also need a nil check for `UsingIpSpace` as it is `*bool`, this may cause a NPE if using ip space is not found as part of the RDE response due to it being omitempty
</blockquote></details>

Addressed.

The length check above will make sure that the uplink is not nil

**Pull request merged**